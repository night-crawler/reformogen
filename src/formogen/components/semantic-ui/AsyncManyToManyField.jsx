import React from 'react';
import { Form } from 'semantic-ui-react';
import Select from 'react-select';
import _ from 'lodash';
import 'react-select/dist/react-select.css';
import fetch from 'isomorphic-fetch';
import Label from './Label';
import propTypes from '../fieldPropTypes';
import * as URI from 'urijs';

import ModelInstanceOption from './ReactSelectOptionComponent';
import ModelInstanceValue from './ReactSelectValueComponent';


AsyncManyToManyField.propTypes = propTypes;
export default function AsyncManyToManyField(props) {
    /*
        "name": "inspire_source",
        "verbose_name": "inspire source",
        "help_text": "",
        "blank": true,
        "null": false,
        "editable": true,
        "type": "ForeignKeyField",
        "required": false,
        "data": []
    */
    const handleChange = (val) => {
        let plainIds = val.map( ({ id }) => (id*1) );
        props.onChange(null, {name: props.name, value: plainIds});
    };

    const loadOptions = (input, callback) => {
        let uri = URI(props.data).addSearch({q: input});

        fetch(uri)
            .then(response => response.json())
            .then(json => {
                if (_.isArray(json)) {  /* without pagination */
                    callback(null, { complete: false, options: json, });
                } else {  /* with pagination */
                    callback(null, { complete: false, options: json.results, });
                }
            });
    };

    let _props = {
        clearable: !props.required,
        closeOnSelect: true,
        disabled: !props.editable,
        multi: true,
        onChange: handleChange,
        placeholder: props.placeholder,

        value: props.value,
        inputProps: {type: 'react-type'},  // fixes broken semantic markup
        removeSelected: true,

        valueKey: 'id',  /* server-side model should provide `id` for object */
        // simpleValue: true,  /* we have no need to store whole objects in parent's state */

        optionComponent: ModelInstanceOption,
        valueComponent: ModelInstanceValue,

        // https://github.com/JedWatson/react-select#advanced-filters
        // we use server-side filters, Select should not filter it again
        filterOption: () => true,
        loadOptions: loadOptions,

        // rtl: this.state.rtl,
    };

    if (_.isFunction(props.updateProps)) {
        _props = props.updateProps(_props, props);
    }

    return (
        <Form.Field required={ props.required } disabled={ !props.editable }>
            <Label { ...props } />
            <Select.Async { ..._props } />
            { !props.helpTextOnHover ? <span className="help-text">{ props.help_text }</span> : ''}
        </Form.Field>
    );
}
