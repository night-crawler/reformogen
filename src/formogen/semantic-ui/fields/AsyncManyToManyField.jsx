import PropTypes from 'prop-types';
import React from 'react';
import { isEmpty, map } from 'lodash';
import { Form } from 'semantic-ui-react';
import AsyncSelect from 'react-select/lib/Async';

import propTypes from '../../fieldPropTypes';
import { FieldLabel } from '../FieldLabel';
import { ErrorsList } from '../ErrorsList';


AsyncManyToManyField.propTypes = {
  ...propTypes,
  getOptionLabel: PropTypes.func,
  getOptionValue: PropTypes.func,
};
AsyncManyToManyField.defaultProps = {
  loadOptions: payload => // eslint-disable-next-line
    console.info(`AsyncManyToManyField.loadOptions(${JSON.stringify(payload)})`),
  loadMoreOptions: payload => // eslint-disable-next-line
    console.info(`AsyncManyToManyField.loadMoreOptions(${JSON.stringify(payload)})`),

  getOptionLabel: ({ name }) => name,
  getOptionValue: ({ id }) => id,
};
export function AsyncManyToManyField(props) {
  const handleChange = val => {
    props.onChange(null, { 
      name: props.name, 
      value: map(val, props.getOptionValue),
    });
  };
    
  let storedCallback = () => {};
  let storedInputText = '';

  return (
    

    <Form.Field
      required={ props.required }
      disabled={ !props.editable }
      width={ props.layoutOpts.width }
      error={ !isEmpty(props.errors) }
    >
      <FieldLabel { ...props } />
      <AsyncSelect
        inputId={ `AsyncManyToManyField-${props.formId}-${props.name}` }
        placeholder={ props.placeholder }
        isClearable={ props.editable }
        isDisabled={ !props.editable }
        isLoading={ props.isLoading }
        isRtl={ props.isRtl }
        isMulti={ true }
        cacheOptions={ true }
        defaultOptions={ true }
        loadOptions={ (inputText, callback) => {
          storedCallback = callback;
          storedInputText = inputText;
          return props.loadOptions({ 
            formId: props.formId,  
            inputText: inputText, 
            fieldName: props.name,
            url: props.data,
            callback: callback,
          });
        } }

        onMenuScrollToBottom={ () => props.loadOptions({
          formId: props.formId,  
          inputText: storedInputText, 
          fieldName: props.name,
          url: props.data,
          callback: storedCallback,
        }) }

        value={ props.value }

        getOptionLabel={ props.getOptionLabel }
        getOptionValue={ props.getOptionValue }

        onChange={ handleChange } 
      />
      { !props.helpTextOnHover
        ? <span className='help-text'>{ props.help_text }</span>
        : ''
      }
      <ErrorsList messages={ props.errors } />
    </Form.Field>
  );
}
