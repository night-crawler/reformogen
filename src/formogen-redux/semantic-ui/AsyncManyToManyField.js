import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { formId, finalFieldValue, initialFieldValue, asyncFieldOptions } from '~/formogen-redux/selectors';

import { storeFieldData } from '~/formogen-redux/actions';

import { AsyncManyToManyField as AsyncManyToManyFieldComponent } from '~/formogen/semantic-ui/fields/AsyncManyToManyField';

function mergeProps(stateProps, dispatchProps, ownProps) {
  const { dispatch } = dispatchProps;
  const { formId } = ownProps;

  return {
    ...ownProps,
    ...stateProps,
    ...dispatchProps,

    onChange: (e, { name, value }) => dispatch(storeFieldData(formId, name, value)),
    dispatch: undefined,
  };
}

export const AsyncManyToManyField = connect(
  createStructuredSelector({
    formId,
    initialValue: initialFieldValue,
    value: finalFieldValue,
    options: asyncFieldOptions
  }),
  dispatch => ({ dispatch }),
  mergeProps,
)(AsyncManyToManyFieldComponent);
