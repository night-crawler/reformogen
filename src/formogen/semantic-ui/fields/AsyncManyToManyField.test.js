import { AsyncManyToManyField } from './AsyncManyToManyField';


it('<AsyncManyToManyField />', () => {
  const wrapper = shallow(
    <AsyncManyToManyField
      type='AsyncManyToManyField'
      name='AsyncManyToManyField: name'
      verbose_name='AsyncManyToManyField: verbose_name'
      help_text='AsyncManyToManyField: help_text'
      displayOptions={ { width: 1 } }
    />
  );
  expect(wrapper).toMatchSnapshot();
});
