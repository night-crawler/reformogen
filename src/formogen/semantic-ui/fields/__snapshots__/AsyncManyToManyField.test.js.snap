// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AsyncManyToManyField /> 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AsyncManyToManyField
    help_text="AsyncManyToManyField: help_text"
    layoutOpts={
      Object {
        "width": 1,
      }
    }
    name="AsyncManyToManyField: name"
    type="AsyncManyToManyField"
    verbose_name="AsyncManyToManyField: verbose_name"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "getOptionLabel": [Function],
      "getOptionValue": [Function],
      "help_text": "AsyncManyToManyField: help_text",
      "isMulti": true,
      "layoutOpts": Object {
        "width": 1,
      },
      "loadOptions": [Function],
      "name": "AsyncManyToManyField: name",
      "type": "AsyncManyToManyField",
      "verbose_name": "AsyncManyToManyField: verbose_name",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "getOptionLabel": [Function],
        "getOptionValue": [Function],
        "help_text": "AsyncManyToManyField: help_text",
        "isMulti": true,
        "layoutOpts": Object {
          "width": 1,
        },
        "loadOptions": [Function],
        "name": "AsyncManyToManyField: name",
        "type": "AsyncManyToManyField",
        "verbose_name": "AsyncManyToManyField: verbose_name",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
